### ISSUE TITLE
Replace IntEnum trait with Enumeration schema tag

### ISSUE THREAD

**Issue by kubukoz** [disneystreaming/smithy4s#932]
In a similar fashion to primitives and collections, we should offer a tag that'll make it more obvious that `intEnum` also needs to be handled.

Instead of:

```scala
def enumeration[E](shapeId: ShapeId, hints: Hints, values: List[EnumValue[E]], total: E => EnumValue[E]): F[E]
```

we'd have

```scala
def enumeration[E](shapeId: ShapeId, hints: Hints, values: List[EnumValue[E]], total: E => EnumValue[E], tag: EnumTag): F[E]
```

where `EnumTag` would be either `StringEnum extends EnumTag` or `IntEnum extends EnumTag`.

This should also be reflected in the `EnumerationSchema` type, as well as the code generator.

**Comment by Baccata** [comment]
I've initially encoded it as in as an `hint`, because I fundamentally disagree with AWS's notion that distinction should be part of the metamodel. 

That being said I can agree that it's not ideal for visibility and makes it easy to overlook the case of int enums. Let's roll with a simple ADT tag then.

**Comment by kubukoz** [comment]
Do we change codegen too? That'd make it consistent with other such tags. I'm fine just having it in the visitor, just wanted to see your take on it

**Comment by Baccata** [comment]
Yeah I think you'd have to change the `Schema` model really, and therefore the codegen. I mean, you could create a short-hand `stringEnum` and `intEnum` smart constructors in Schema, but you'll have to change the codegen in consequence no matter what.

**Comment by yisraelU** [comment]
Does this mean , we would remove it from Hints  ?

**Comment by Baccata** [comment]
@yisraelU probably indeed.

**Comment by kubukoz** [comment]
yeah, that's what I'm thinking.

speaking of which, I'd kinda like to get rid of `InputOutput`... but it's a can of worms I don't want to open right now

**Comment by Baccata** [comment]
> speaking of which, I'd kinda like to get rid of InputOutput... but it's a can of worms I don't want to open right now

@kubukoz we can and should, now that we have a way of filter schemas and a more granular mechanism than `CodecAPI`.

**Comment by kubukoz** [comment]
well that was easy :)

**Comment by kubukoz** [comment]
Created a ticket internally for my team, somebody will probably pick this up soon (I'll then reassign)

**Comment by kubukoz** [comment]
I can't assign him (presumably due to not having interacted with this repo before) but @GaryAghedo will be on it :)

**Comment by miguel-vila** [comment]
addressed at #952, think this can be closed
